lnrpc/routerrpc/router_server.go

```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+	// sats, then 1010 sats will have to arrive at the last hop. This means
+	// that the probe has to be dispatched with 1010 sats to correctly
+	// calculate the routing fee.
+	//
+	// Calculate the hop fee for the last hop manually.
+	hopFee := lspHopHint.HopFee(int64(*payReq.MilliSat))
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += hopFee
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to the
+	// to fee the payment probe returned.
+	resp.RoutingFeeMsat += hopFee
+
+	return resp, nil
+}
+
+// isLSP checks if the route hints indicate an LSP. An LSP is indicated in the
+// first return value if the last node in each route hint has the same node id.
+// If an LSP is indicated, the remaining hops in each route hint are returned
+// to the caller to be considered by the caller as route hints to the LSP node.
+// Additionally, the identified LSP node is returned as a separate hop hint.
+// This hop hint also contains the maximum fee setting if LSP hop hints contain
+// different fee settings. This is to obtain an upper bound of the fee estimate.
+func isLSP(routeHints [][]zpay32.HopHint) (bool, [][]zpay32.HopHint,
+	*zpay32.HopHint) {
+
+	if len(routeHints) == 0 || len(routeHints[0]) == 0 {
+		return false, [][]zpay32.HopHint{}, nil
+	}
+
+	// We are likely dealing with an LSP. Hence, we construct a modified
+	// list of route hints that allows the caller to probe the LSP, which
+	// itself is returned as a separate hop hint.
+	adjustedRouteHints := make([][]zpay32.HopHint, 0, len(routeHints))
+
+	// If the final pubkey of each route hint is the same, we can assume
+	// that this identifies an LSP node. The first candidate is our
+	// reference hop hint.
+	refHopHint := routeHints[0][len(routeHints[0])-1]
+
+	// Strip off the potential LSP hop hint from the reference route hint,
+	// and add the remaining hop hints to the adjusted route hints.
+	if len(routeHints[0]) > 1 {
+		adjustedRouteHints = append(
+			adjustedRouteHints,
+			routeHints[0][:len(routeHints[0])-1],
+		)
+	}
+	for i := 1; i < len(routeHints); i++ {
+		rh := routeHints[i]
+
+		// Skip empty route hints.
+		if len(rh) == 0 {
+			continue
+		}
+
+		// If the last hop hint is not the same as the reference hop
+		// hint, we might not be dealing with an LSP.
+		rhLastHop := rh[len(rh)-1]
+		idMatchesRefNode := bytes.Equal(
+			rhLastHop.NodeID.SerializeCompressed(),
+			refHopHint.NodeID.SerializeCompressed(),
+		)
+		if !idMatchesRefNode {
+			return false, [][]zpay32.HopHint{}, nil
+		}
+
+		// We are interested in an upper bound of a fee estimate when
+		// paying an LSP invoice. If the LSP hop hints indicate
+		// different final hop fees we track the highest.
+		if refHopHint.HopFee(1_000_000) < rhLastHop.HopFee(1_000_000) {
```

Comment:
this could use the payment amount from the invoice to make it precise

Reply:
nice!